src/ai/backend/agent/server.py

```diff
@@ -500,18 +513,52 @@ async def _pull(reporter: ProgressReporter, *, img_conf: ImageConfig) -> None:
                 img_ref, img_conf["digest"], AutoPullBehavior(img_conf["auto_pull"])
             )
             if need_to_pull:
+                log.info(f"rpc::check_and_pull() start pulling {str(img_ref)}")
                 await self.agent.produce_event(
-                    ImagePullStartedEvent(image=str(img_ref), agent_id=self.agent.id)
+                    ImagePullStartedEvent(
+                        image=str(img_ref),
+                        agent_id=self.agent.id,
+                    )
                 )
                 image_pull_timeout = cast(
                     Optional[float], self.local_config["agent"]["api"]["pull-timeout"]
                 )
-                await self.agent.pull_image(
-                    img_ref, img_conf["registry"], timeout=image_pull_timeout
+                try:
+                    await self.agent.pull_image(
+                        img_ref, img_conf["registry"], timeout=image_pull_timeout
+                    )
+                except asyncio.TimeoutError:
+                    log.exception(f"Image pull timeout after {image_pull_timeout} sec")
+                    await self.agent.produce_event(
+                        ImagePullFailedEvent(
+                            image=str(img_ref),
+                            agent_id=self.agent.id,
+                            msg=f"timeout (s:{image_pull_timeout})",
+                        )
+                    )
+                except Exception as e:
+                    log.exception(f"Image pull failed (e:{repr(e)})")
+                    await self.agent.produce_event(
+                        ImagePullFailedEvent(
+                            image=str(img_ref),
+                            agent_id=self.agent.id,
+                            msg=repr(e),
+                        )
+                    )
+                else:
+                    await self.agent.produce_event(
+                        ImagePullFinishedEvent(
+                            image=str(img_ref),
+                            agent_id=self.agent.id,
+                        )
+                    )
+            else:
+                await self.agent.produce_event(
+                    ImagePullFinishedEvent(
+                        image=str(img_ref),
+                        agent_id=self.agent.id,
+                    )
                 )
```

Comment:
There seems to be a situation where a finish event can be delivered without a started event, is this okay where the event is being handled?
It does seem like there is an inconsistency in producing events.

Reply:
yes, this API can produce "finish event" without "start event". Would it be better to have "noop event" besides "finish event"?